hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java

```diff
@@ -955,11 +955,27 @@ private void populateBuffer(int length,
                               S3ObjectInputStream objectContent) throws IOException {
     if (buffer.isDirect()) {
       byte[] tmp = new byte[length];
-      objectContent.read(tmp, 0, length);
+      readByteArray(objectContent, tmp, 0, length);
```

Comment:
maybe we should create a buffer of a max size, read in blocks and put incrementally. otherwise we alloc a 1GB array to load a range that big, when we don't need so much heap

Reply:
I made this change but then realised, the temp array will never be greater than maxReadSizeForVectorReads which is by default 1 MB.